src/search.cpp

```diff
@@ -211,18 +210,62 @@ void Search::clear() {
   Tablebases::init(Options["SyzygyPath"]); // Free mapped files
 }
 
+/// MainThread::perft() divide the perft search evenly through Threads
+void MainThread::perft() {
+
+  StateInfo st;
+  perftDepth = Limits.perft;
+  std::string posNow = rootPos.fen();
+  std::vector<PerftPos> newPerftRoot;
+  perftIdx.store(0, std::memory_order_release);
+
+  // Step 1: Create new root positions that will be distributed between threads
+  for (const auto& m : MoveList<LEGAL>(rootPos)) {
+      rootPos.do_move(m, st);
+      perftSum[perftRoot.size()].store(0, std::memory_order_release);
+      perftRoot.push_back({rootPos.fen(), (int)perftRoot.size()});
+      rootPos.undo_move(m);
+  }
+
+  // Step 2: Extend root positions if it's not enough to distribute evenly between threads
+  while (--perftDepth > 3 && 8 * Threads.size() > perftRoot.size()){
+      newPerftRoot.clear();
+      for (const auto& pr : perftRoot){
+          rootPos.set(pr.fen, rootPos.is_chess960(), rootPos.state(), this);
+
+          for (const auto& m : MoveList<LEGAL>(rootPos)){
+              rootPos.do_move(m, st);
+              newPerftRoot.push_back({rootPos.fen(), pr.ID});
+              rootPos.undo_move(m);
+          }
+      }
+      perftRoot = newPerftRoot;
+  }
+
+  // Step 3: Initialize Perft in all threads and wait all to be finished
```

Comment:
Maybe it's redundant or obvious, but it's good to remark the following IMHO:
`// Note that search threads will view the latest state of perftRoot, thanks to internal mutex synchronization during the search start.`

Reply:
what do you think about `perftRoot and perftDepth are updated in all Threads due to internal mutex synchronization.`? Does it explain what is going on? I don't want to overexplain.

Reply:
Actually "updated" is a bit weird, as this seems to imply they are somehow copies, whereas they're just shared variables. Maybe "consistent"?